src/test/java/com/uber/cadence/workflow/LoggerTest.java

```diff
@@ -118,7 +120,8 @@ public void testWorkflowLogger() {
 
   private int matchingLines(String message) {
     int i = 0;
-    for (ILoggingEvent event : listAppender.list) {
+    List<ILoggingEvent> loggingEvents = new ArrayList<>(listAppender.list);
+    for (ILoggingEvent event : loggingEvents) {
```

Comment:
seems unrelated, the list is still being iterated over to make the copy -> that can still race with something modifying the list.

Reply:
this technique is called []defensive copying](http://www.javapractices.com/topic/TopicAction.do?Id=15). _ArrayList_ has a _copying constructor_ so when the original list (from appender) get modified it does not affect the copy so the ConcurrentModificationException is avoided.

Reply:
I am quite bad at keeping up with github notifications, so apologies about the (gigantic) delay on this :\  it was not intended, I didn't notice the response and then it was gone forever.

Defensive copying does not apply in this kind of concurrent modification - what's happening here is that _other threads_ are modifying the list while it's being iterated - that iteration is still happening with this copy, it's just happening inside the ArrayList constructor, so the race is still here.  It has just moved one line earlier.

Defensive copying only solves the case where the contents of the loop modify the collection being looped over.  For that scenario it's useful / a real fix, but that's not occurring here.

#885 has a correct fix: synchronize access to the list itself.  
Ideally we wouldn't have background threads pushing data in here after the test has "ended", but that's unfortunately likely to take a lot more work to solve :\